### PR TITLE
[Feature][MRV2] Optimize resample_kernel with 150x speed up

### PR DESCRIPTION
**Pull request by AuroraEmiya**
### What this PR does / why we need it?

The current MRV2 rejection resampling path on Ascend uses Gumbel-Max sampling, which generates Gumbel noise for every vocabulary token and performs block-local argmax reductions before selecting the final token.

This PR replaces the Gumbel-Max resampling stage with a direct Triton categorical sampling implementation.

The new implementation:

* computes block-level probability mass instead of full-vocabulary Gumbel noise;
* supports full-draft residual sampling with `max(p - q, 0)`;
* supports one-hot draft residual sampling by excluding the rejected draft token;
* preserves target softmax sampling for bonus tokens;
* preserves the existing greedy rejection and greedy bonus semantics;
* uses one random value per request and propagates one global categorical threshold through block-level and token-level selection;
* aligns `_resample_kernel` with the available Ascend Vector Cores and evenly distributes `(request, vocabulary block)` tasks across workers;
* keeps probability-mass computation in FP32 while supporting FP16, BF16, and FP32 logits;
* moves resampling into a standalone Triton `resample()` operator and updates the existing MRV2 resample tests and documentation.

The rejection verification logic itself is unchanged.

For the profiled `num_reqs=8, vocab_size=151936` case, `msprof op` reports:

```text
_resample_kernel
Before core-aligned task mapping:
  Task Duration: 155.38 us
  Block Dim:     1192

After core-aligned task mapping:
  Task Duration: 33.84 us
  Block Dim:     40

_categorical_finalize_kernel
  Task Duration: 27.04 us
  Block Dim:     8
```

The Vector-Core-aligned mapping provides about a **4.59x speedup** for the categorical `_resample_kernel` itself.

### Does this PR introduce *any* user-facing change?

Yes, but there is no public API or distributional semantic change.

The resampling algorithm changes from Gumbel-Max to direct categorical sampling. Both sample from the same target categorical distribution, but the exact sampled token stream for the same seed and position may differ because the random-number mapping is different.

The `rejection_sample` interface, accepted-token semantics, residual distribution, bonus-token distribution, and greedy behavior remain unchanged.

### How was this patch tested?

The existing MRV2 resample numerical test and operator documentation are updated for the categorical implementation.

The tests exercise the public `resample()` API using independent PyTorch or analytic references rather than reproducing the Triton implementation.

Coverage includes:

* finite logits;
* partially masked `-inf` logits;
* vocabulary tails around the 1024-token block boundary;
* full-draft `(p - q)+` residual sampling;
* one-hot draft residual sampling;
* random bonus softmax sampling;
* greedy bonus and greedy rejection;
* FP16, BF16, and FP32 logits;
* ragged request lengths;
* shuffled request-state mappings;
* mixed greedy/random and bonus/rejection requests;
* deterministic seed/position behavior;
* `num_reqs=8, vocab_size=151936`;
* end-to-end greedy `rejection_sample` integration.

The statistical cases compare observed sampling frequencies against the analytic softmax or normalized `(p - q)+` distributions.

```bash
pytest -sv tests/e2e/nightly/single_node/ops/singlecard_ops/triton/test_resample.py
```

Performance was profiled with `msprof op` for `_resample_kernel` and `_categorical_finalize_kernel`.

* vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

**Review by Liamup777 — COMMENTED** [review]
<html><head></head><body><h3>Performance</h3><p>Tested <code inline="">_resample_kernel</code> with the real workload in the end-to-end model.</p>

  | Count | Total Time (us) | Min Time (us) | Avg Time (us) | Max Time (us) | Ratio
-- | -- | -- | -- | -- | -- | --
Before | 51 | 2706913.207 | 1256.566 | 53076.729 | 74675.512 | 24.959%
After | 51 | 12873.54 | 49.06 | 252.421 | 351.84 | 0.16%

<p>The average kernel latency is reduced from <strong>53076.729 us to 252.421 us</strong>, achieving about <strong>210x speedup</strong>.</p><p>The kernel's end-to-end execution time ratio is reduced from <strong>24.959% to 0.16%</strong>.</p></body></html>

**Review by chenlongxiao — APPROVED** [review]

**Review by zzzzwwjj — APPROVED** [review]
